internal/handler/cmd_query.go

```diff
@@ -28,19 +29,30 @@ import (
 
 // CmdQuery implements deprecated OP_QUERY message handling.
 func (h *Handler) CmdQuery(ctx context.Context, query *wire.OpQuery) (*wire.OpReply, error) {
-	cmd := query.Query().Command()
+	body := query.Query()
+	cmd := body.Command()
 	collection := query.FullCollectionName
 
 	// both are valid and are allowed to be run against any database as we don't support authorization yet
 	if (cmd == "ismaster" || cmd == "isMaster") && strings.HasSuffix(collection, ".$cmd") {
-		return common.IsMaster(ctx, query.Query(), h.TCPHost, h.ReplSetName)
+		return common.IsMaster(ctx, body, h.TCPHost, h.ReplSetName)
 	}
 
 	// TODO https://github.com/FerretDB/FerretDB/issues/3008
 
 	// database name typically is either "$external" or "admin"
 
 	if cmd == "saslStart" && strings.HasSuffix(collection, ".$cmd") {
+		mechanism, err := common.GetRequiredParam[string](body, "mechanism")
+		if err == nil && mechanism == "PLAIN" {
```

Comment:
Let's return the error for the case where `err != nil`. :hugs: 

Let's also handle the case where `mechanism != "PLAIN"`. I think we can handle it similarly to how it's done in `msg_saslstart.go`

Reply:
The idea was to provide extreme backwards compatibility by just falling back to the "stub" behavior if something went wrong, as I have no ideas if that would break any other clients. But I don't mind make explicit errors here...

Reply:
Thanks for your quick response :hugs:. Let us check, @b1ron do you know which other driver uses `cmd_query.go`? My guess is, it doesn't work for those drivers? If that's the case having an appropriate error is sensible here for a client attempting to authenticate with other mechanisms.

Reply:
It depends on the specific driver and version. Looking at this [compatibility matrix](https://www.mongodb.com/docs/drivers/csharp/current/compatibility/#compatibility-table-legend) drivers that do not support MongoDB 6.0 will use legacy opcodes. I suspect @Evengard is using mongocsharpdriver version 2.15 or older.

> OP_QUERY
Deprecated in MongoDB 5.0. Removed in MongoDB 5.1.

Fixing #3121 might be easier.

Reply:
> Removed in MongoDB 5.1.

*Narrator: It wasn't actually removed.*

But yeah, to move forward, we have to know the driver version.

Reply:
![изображение](https://github.com/FerretDB/FerretDB/assets/798735/7b8ccbbe-8422-49f6-b3b9-fdfad102e835)

I specifically made sure to upgrade to the latest revision. This is the behaviour with latest stable.

Reply:
https://github.com/mongodb/mongo-csharp-driver/blob/20391c999873faf71e96e5109e2fda00a84a6e9a/src/MongoDB.Driver.Core/Core/WireProtocol/CommandWireProtocol.cs#L181

There's that method which seems to fallback to the OP_QUERY if it deemed the server version is too low. Maybe you have some pointers about how to force it use OP_MSG instead?

Reply:
I've actually found that the commit that fixes that behaviour in the C# client is there: https://github.com/mongodb/mongo-csharp-driver/commit/7be56f72ff484ffdeb9f021a8d582b4b6d62eeb8 except it is really recent, it didn't make yet into an official MongoDB.Driver release. I tried to compile the main branch of the MongoDB.Driver and tried to connect with it without this PR, and it actually worked. Not with the latest official 2.23.1 though.

Reply:
Looks like it is nearing the release: https://github.com/mongodb/mongo-csharp-driver/pull/1261

Reply:
I was able to force an OP_MSG using version 2.19.1, strange.

Reply:
The thing is I can force an OP_MSG authentication with StableAPI, it's the inserts which bugs then.

I can do a PR in attempts to ignore theese apiVersions parameters probably instead. But looking that the actual issue seemed abandoned wasn't sure it is the way to go.

Reply:
https://github.com/FerretDB/FerretDB/pull/4067 Went ahead with the StableAPI parameter ignoring... It works with 2.23.1 too.

Reply:
Thanks @Evengard, and apologies for my confusion. I was able to reproduce the error in both 2.19.1 and 2.23.1.

```
Unhandled exception. MongoDB.Driver.MongoCommandException: Command ping failed: [msg_ping.go:52 handler.(*Handler).MsgPing] [backend.go:67 postgresql.(*backend).Status] [registry.go:259 metadata.(*Registry).DatabaseList] [registry.go:132 metadata.(*Registry).getPool] [pool.go:128 pool.(*Pool).Get] [opendb.go:95 pool.openDB] [opendb.go:110 pool.checkConnection] failed to connect to `host=postgres user= database=ferretdb`: server error (FATAL: no PostgreSQL user name specified in startup packet (SQLSTATE 28000)).
```
#4067 fixes the issue I also tested it. The core team will review it.

Reply:
We should do https://github.com/FerretDB/dance/issues/769 first